doc/api/cli.md

```diff
@@ -1370,8 +1370,15 @@ added: v12.0.0
 -->
 
 This configures Node.js to interpret `--eval` or `STDIN` input as CommonJS or
-as an ES module. Valid values are `"commonjs"` or `"module"`. The default is
-`"commonjs"`.
+as an ES module. Valid values are `"commonjs"`, `"module"`, `"module-typescript"` and `"commonjs-typescript"`.
+The `"-typescript"` values are available only in combination with the flag `--experimental-strip-types`.
+The default is `"commonjs"`.
+
+If `--experimental-strip-types` is enabled and `--input-type`is not provided,
+Node.js will first try to run the code as JavaScript,
+then it will try to run the code as TypeScript.
```

Comment:
We should probably say something about syntax detection here, that first it attempts to run as CommonJS JavaScript, and depending on the parse error thrown, will retry as ESM JavaScript or CommonJS TypeScript; and then potentially a third time as ESM TypeScript? I’m not even sure what the flow would be, which probably means that we should spell it out 😄

Reply:
I'm not an expert on how the syntax detection is implemented but the flow is:
- run normally as javascript (this will perform the detection by trying cjs and esm)
- if invalid syntax, strip the types and try again
- if invalid again throw the first invalid syntax error enriched with the typescript message

Reply:
The details of syntax detection are here: https://github.com/nodejs/node/blob/6012a4e9a2733b25a7023baa0a58bef85c9ebc5f/doc/api/esm.md#:~:text=at%20pjsonURL.-,DETECT_MODULE_SYNTAX,-(source)

The important part is to clarify which specific errors will trigger strip-types behavior. I assume there’s no overlap between the “retry as ESM” errors and the “retry with type stripping” errors, so the flow is no more expensive than the current detection flow: try first as CommonJS JavaScript, and retry as stripped types.

It gets even more complicated because doesn’t the strip-types flow itself use syntax detection? Like after it strips the types, then it initially tries to run as CommonJS, then retries as ESM?

Reply:
Yes

Reply:
I think the implementation is fine, let’s just improve the documentation to better explain it? Is it this?

1. Try to run as CommonJS JavaScript.
2. If the above fails because of a parse error related to module syntax, retry as ESM JavaScript.
3. If either of the above fails because of a parse error related to types syntax, strip the types and run as stripped CommonJS JavaScript.
4. If the above fails because of a parse error related to module syntax, retry as ESM JavaScript.